wallet: don't read db every time that a new 'WalletBatch' is created

furszy opened the pull request:
Found it while was working on #25297.

We are performing a db read operation every time that a new `WalletBatch` is created, inside the constructor, just to check if the client version field is inside the db or not.

As the client version field does not change in the entire db lifecycle, this operation can be done only once: The first time that the db is accessed/opened and the client version value can be cached.

furszy commented on src/wallet/bdb.cpp:
```diff
@@ -315,11 +315,17 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const bool read_only, b
-        fReadOnly = fTmp;
+
+    // At the first db open
+    if (database.client_version == -1) {
```
Mainly because of how the flow is currently structured. We are actually checking if the db is open on every new batch, and opening it if not (that is done few lines above, at line 312).

In other words and to be a bit more specific: I was going to add it inside the `BerkeleyDatabase::Open()` method but.. as `BerkeleyBatch` contains the functions to access the db (Read/Write/Erase etc), I was going to end up duplicating the code there (the `BerkeleyBatch::ReadKey` functionality).

Not sure if I should include it here but I think that a good next step would be to decouple the db.open call from the `BerkeleyBatch` constructor, so it only receives an already opened database. And in that way, the client version read could be decoupled as well and not be part of the constructor.

laanwj replied:
Fair enough. If you have a good reason for it it's okay with me. I was just asking.
It's "only" the legacy wallet so doing any large refactors is probably not worth it.

furszy replied:
yeah, no problem. The question was great to be sync about following-up improvements. The SQLite flow has some similarities.. (the db constructor is, mostly, a no-op as the db is always opened etc..).

plus, I'm grateful for your quick feedback :).

w0xlt reviewed (commented):
Approach ACK

achow101 commented:
Honestly this check doesn't even need to exist. I can't quite figure out why it was added in the first place as the code is actually traceable to the first commit in the project! I think it's related to something with using bdb as the transaction index back then. Nowadays, this record is only used to record the most recent client that has opened the wallet, and it is only relevant if the wallet was opened by 0.4 or 0.5 which had database encryption issues.

As the version record is updated in `LoadWallet`, I think it is safe to remove this code and just update `LoadWallet` to write the record if it doesn't already exist.

furszy committed:
wallet: don't read db every time that a new WalletBatch is created

Better to perform the action only one time (during 'LoadWallet').
Where the value is being used.

furszy commented:
Pushed. Ready to go.

w0xlt reviewed (approved):
Code Review ACK https://github.com/bitcoin/bitcoin/pull/25383/commits/80400e9bbe9e7af72acab142a2327511e8b0d036

MarcoFalke commented on src/wallet/walletdb.cpp:
```diff
@@ -909,7 +909,7 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         return DBErrors::NEED_REWRITE;
 
-    if (last_client < CLIENT_VERSION) // Update
+    if (!has_last_client || last_client < CLIENT_VERSION) // Update
```
Any reason to keep this check? If yes, maybe rename `last_client` to `highest_client`?

furszy replied:
> Any reason to keep this check? 

The purpose of this process is store (first condition) or update (second condition) the version of the last client who opened the wallet db. Plus, we don't want to write to db if the value is already there.

> maybe rename last_client to highest_client?

hmm, I don't think that `highest_client` represents what this variable really is. I mean, the comment that is above the variable declaration is saying "// Last client version to open this wallet, was previously the file version number".

Which doesn't necessarily need to be the "highest" version.

MarcoFalke replied:
Yeah, the comment seems wrong as well. If you open with Bitcoin Core 23.0, it will write the 23.0 CLIENT_VERSION. If you then open with 22.0, it will not update because `last_client < CLIENT_VERSION` evaluates to false?

furszy replied:
oh yeah, that is also wrong. Updating.. :)

achow101 replied:
Hmm, indeed, that is a misnomer. The comment and variable name should be changed to reflect what it actually does.

furszy replied:
Hmm, I was going into the opposite direction actually: update the client version value if it's different than the new version. So we always know (and can log) which soft version opened the wallet db last (something like this https://github.com/bitcoin/bitcoin/pull/25383/commits/c318211ddd48d44dd81dded553afeee3bc41c89e).

MarcoFalke commented on src/wallet/walletdb.cpp:
```diff
@@ -885,7 +885,7 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
+    bool has_last_client = m_batch->Read(DBKeys::VERSION, last_client);
 
     int wallet_version = pwallet->GetVersion();
     pwallet->WalletLogPrintf("Wallet File Version = %d\n", wallet_version > 0 ? wallet_version : last_client);
```
unrelated: Any reason for the `> 0` check? If the minimum version needed to read the wallet is `0`, it should probably be printed as-is?

furszy replied:
yeah, that doesn't make sense. Even `0.4.0` wallet version was `40000`.

furszy commented:
Thanks for the feedback!

Squashed https://github.com/bitcoin/bitcoin/pull/25383#discussion_r899149943 inside bda8ebe.

Now we will log both versions; the wallet version and the version of the last client who opened the db.

furszy committed:
walletdb: fix last client version update

The value was only being updated launching releases with higher version numbers
and not if the user launched a previous release.

Co-authored-by: MacroFake <falke.marco@gmail.com>

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/25383/commits/c318211ddd48d44dd81dded553afeee3bc41c89e

The change since my last ACK makes sense if the purpose of `last_client` and `CLIENT_VERSION` is to store the last client that opened the database and not the highest, as suggested by the original code.

luke-jr commented on src/wallet/walletdb.cpp:
```diff
@@ -909,7 +907,7 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         return DBErrors::NEED_REWRITE;
 
-    if (last_client < CLIENT_VERSION) // Update
+    if (!has_last_client || last_client != CLIENT_VERSION) // Update
```
Looks like this is changing the semantics of "last client". It used to be the newest version ever used, but now it is the most recent version used, newer or not. I'm not sure the latter is preferable - and even if so, maybe it should use a new key.

NawaraTafa replied:
كيف اقوم بسحب الاموال

furszy replied:
> Looks like this is changing the semantics of "last client". It used to be the newest version ever used, but now it is the most recent version used, newer or not. I'm not sure the latter is preferable

yeah, not sure if you have checked it already but we had the discussion above: https://github.com/bitcoin/bitcoin/pull/25383#discussion_r899188122.

Comments there are very welcome, I'm fine either way.
I moved forward with what seemed to be the desired workflow for this variable (based on the code comments). At the end, as achow101 said, this variable is never actually used for anything important except for the 0.4 and 0.5 encryption fixes (on which, this will still be fine).

achow101 replied:
@luke-jr I think the new behavior is actually an improvement and more correct than the previous. Arguably it's a bug fix.

Suppose a user creates a wallet in a version more recent than 0.5. They then downgrade to 0.4 and load that wallet and encrypt it there. They then open the wallet again in a newer version.

What we want to happen is that opening the wallet in the newer version will force a rewrite of the wallet because it was encrypted in 0.4 and has the bug where unencrypted keys are still in the database. However, because `version` is the most recent client that has opened the wallet, this does not happen and the wallet file contains the unencrypted keys perpetually. If 0.4 had the behavior implemented here, the `version` record would indicate that 0.4 had last opened the file and so the newer version would correctly do the database rewrite.

While this is a rather contrived scenario, if we do have a similar problem in the future, having the record be written
